data/transactions/logic/opcodes.go

```diff
@@ -319,6 +319,9 @@ var OpSpecs = []OpSpec{
 	{0x96, "bsqrt", opBytesSqrt, asmDefault, disDefault, oneBytes, oneBytes, 6, modeAny, costly(40)},
 	{0x97, "divw", opDivw, asmDefault, disDefault, twoInts.plus(oneInt), oneInt, 6, modeAny, opDefault},
 	{0x98, "sha3_256", opSHA3_256, asmDefault, disDefault, oneBytes, oneBytes, 7, modeAny, costly(130)},
+	{0x99, "bn256_add", opBn256Add, asmDefault, disDefault, twoBytes, oneBytes, 6, modeAny, costly(166)},
+	{0x9a, "bn256_scalar_mul", opBn256ScalarMul, asmDefault, disDefault, twoBytes, oneBytes, 6, modeAny, costly(892)},
+	{0x9b, "bn256_pairing", opBn256Pairing, asmDefault, disDefault, twoBytes, oneBytes, 6, modeAny, costly(6879)},
```

Comment:
Can you comment on how you chose the set of opcodes you feel are complete enough to prove snarks.  In particular, it looks like Ethereum is adding 9 pre-compiles in support of BLS12-381. https://eips.ethereum.org/EIPS/eip-2537

Reply:
Though I see that they only added the three you chose for support of bn256. So I suppose there are good reasons, and the curves are typically used differently.

Reply:
The major reason is cost. bls-381 is more expensive than bn256. And even bn256 barely fit in AVM's 11200 (stateful 16 grouped tx) / 20000 (stateless) cost limit. 

I made mistake in unmarshal, which makes 6879 above wrong. Correct one is about 8000-9000 for pairing one pair. In verifying SNARK it's pairing 4 pairs, which costs about 20000.

**Is there plans to allow group more than 16 txns?** The 11200 cost limit is too strict for bn256 pairing operations. Actually, go-ethereum has selected a less efficient library and got a much more expensive cost.Our bn256 pairing is about 6x (!) faster than go-ethereum, but it's still doesn't fit 11200. This indicates two things:
1) it's unlikely to make non-trivial optimization on the timing over the library that we used for bn256 pairing (gnark's bn254);
2) Ethereum has a much higher gas limit than AVM, if compare the cost by doing same operations
So to support operations as expensive as pairing, I think we do need a higher bar. And theoratically, I think it doesn't hurt to allow expensive operations by paying more tx fee. More specifically, I know a tx can call 256 inner txns to have a totally 179.2K cost limit, so is it possible to group 256 txns to have 179.2K cost limit? :thinking: 

(!): 6x is from both this: https://hackmd.io/@gnark/eccbench and my local benchmark of go-algorand and go-ethereum git-blame of cost parameter, which refer to [eip-1108](https://eips.ethereum.org/EIPS/eip-1108.) and this [bench](https://gist.github.com/zac-williamson/838410a3da179d47d31b25b586c15e53). 
More detail result:

go-ethereum:
```
bn256 pairing: 
3037 microseconds, 1 pairing
14663 microseconds, 10 pairings
= 1745 microseconds + 1292 microseconds per pairing
= 45000 + k*35000 eth gas
when k = 4, it's about 185000 eth gas, 61x 116 ~ 7000 microseconds

In go-algorand
pairing k=1 about 460 microseconds
pairing k=4 about 1000 microseconds
ecrecover on my machine: about 100 microseconds, 2000 cost
```

Reply:
I think you're answering a different question. While opcode cost matters, I'm asking a different question.

Why is it that you (and Ethereum, so you're good company!) have only added three new opcodes to provide support for bn256, but nine new opcodes have been proposed to support bn12-381? I'm trying to ensure we get the basic primitives right, and I'm surprised that two different curves have a different sent of operations.

Reply:
Sorry didn't address your question. **My understanding is 1) 3-primitive in bn256 is sufficient for SNARK verification; 2) BLS curves also have a signature verification use case, which need additional primitives.** Evidence below.

1. Three bn256 primitives are sufficient to verify SNARKs generated from a bn256 curve, [solidity example](https://github.com/ConsenSys/gnark/blob/cc10c32452124bbf2f1115b3d6ac9a05710dc3ab/internal/backend/bn254/groth16/solidity.go) and [ported pyteal](https://github.com/AbstrLabs/algorand_zkp_verifier/blob/verifier-contract-v1/assets/zkp_verifier.py). For same reason we also only added three bn256 primitives in NEAR.

2. Bn256 were added to Ethereum earlier than bls12-381 in [eip-196](https://eips.ethereum.org/EIPs/EIPS/eip-196) and [eip-197](https://eips.ethereum.org/EIPs/EIPS/eip-197). The purpose as mentioned in eip is "perform zkSNARK verification". Bls12-381 were introduced in [eip-2537](https://eips.ethereum.org/EIPS/eip-2537), the purpose are "BLS signature verification and perform SNARKs verifications".

Reply:
updated two broken links